src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -386,6 +392,131 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
     );
   };
 
+  /**
+   * Applies coloring to point traces in a scatterplot. Does this by splitting the data into multiple traces each with a solid
+   * color, which is much faster than using Plotly's native color ramping. Also enforces a maximum number of points
+   * per trace, which significantly speeds up Plotly renders.
+   *
+   * @param xData
+   * @param yData
+   * @param objectIds
+   * @param trackIds
+   * @param markerConfig Additional marker configuration to apply to all points. By default,
+   * markers are size 4.
+   * @param {Color | undefined} overrideColor When defined, uses a base color for all points, instead of
+   * calculating based on the color ramp or palette.
+   */
+  const colorizeScatterplotPoints = (
```

Comment:
I wonder if most or all of this function could move into scatter_plot_data_utils.

Reply:
It could, it currently accesses a lot of state variables that I'd need to pass into it though. Let me think on this one.